snow/consensus/snowman/metrics.go

```diff
@@ -0,0 +1,261 @@
+		),
+		blockSizeAcceptedSum: prometheus.NewGauge(prometheus.GaugeOpts{
+			Namespace: namespace,
+			Name:      "blks_accepted_container_size_sum",
+			Help:      "cumulative size of all accepted blocks",
+		}),
+
+		pollsRejected: metric.NewAveragerWithErrs(
+			namespace,
+			"blks_polls_rejected",
+			"number of polls from the issuance of a block to its rejection",
+			reg,
+			&errs,
+		),
+		// e.g.,
+		// "avalanche_P_blks_rejected_count" to count how many "Observe" gets called -- count all "Reject"
+		// "avalanche_P_blks_rejected_sum" to count how many ns have elapsed since its issuance on rejection
+		// "avalanche_P_blks_accepted_sum / avalanche_P_blks_accepted_count" is the average block acceptance latency in ns
+		// "avalanche_P_blks_accepted_container_size_sum" to track cumulative sum of all accepted blocks' sizes
+		// "avalanche_P_blks_accepted_container_size_sum / avalanche_P_blks_accepted_count" is the average block size
+		latRejected: metric.NewAveragerWithErrs(
+			namespace,
+			"blks_rejected",
+			"time (in ns) from the issuance of a block to its rejection",
+			reg,
+			&errs,
+		),
+		blockSizeRejectedSum: prometheus.NewGauge(prometheus.GaugeOpts{
+			Namespace: namespace,
+			Name:      "blks_rejected_container_size_sum",
+			Help:      "cumulative size of all rejected blocks",
+		}),
+
+		numSuccessfulPolls: prometheus.NewCounter(prometheus.CounterOpts{
+			Namespace: namespace,
+			Name:      "polls_successful",
+			Help:      "number of successful polls",
+		}),
+		numFailedPolls: prometheus.NewCounter(prometheus.CounterOpts{
+			Namespace: namespace,
+			Name:      "polls_failed",
+			Help:      "number of failed polls",
+		}),
+	}
+
+	// Initially set the metrics for the last accepted block.
+	m.maxVerifiedHeight.Set(float64(lastAcceptedHeight))
+	m.lastAcceptedHeight.Set(float64(lastAcceptedHeight))
+	m.lastAcceptedTimestamp.Set(float64(lastAcceptedTime.Unix()))
+
+	errs.Add(
+		reg.Register(m.maxVerifiedHeight),
+		reg.Register(m.lastAcceptedHeight),
+		reg.Register(m.lastAcceptedTimestamp),
+		reg.Register(m.numProcessing),
+		reg.Register(m.blockSizeAcceptedSum),
+		reg.Register(m.blockSizeRejectedSum),
+		reg.Register(m.numSuccessfulPolls),
+		reg.Register(m.numFailedPolls),
+	)
+	return m, errs.Err
+}
+
+func (m *metrics) Issued(blkID ids.ID, pollNumber uint64) {
+	m.processingBlocks.Put(blkID, processingStart{
+		time:       time.Now(),
+		pollNumber: pollNumber,
+	})
+	m.numProcessing.Inc()
+}
+
+func (m *metrics) Verified(height uint64) {
+	m.currentMaxVerifiedHeight = math.Max(m.currentMaxVerifiedHeight, height)
+	m.maxVerifiedHeight.Set(float64(m.currentMaxVerifiedHeight))
+}
+
+func (m *metrics) Accepted(
+	blkID ids.ID,
+	height uint64,
+	timestamp time.Time,
+	pollNumber uint64,
+	blockSize int,
+) {
+	start, ok := m.processingBlocks.Get(blkID)
+	if !ok {
+		m.log.Warn("unable to measure latency",
+			zap.Stringer("blkID", blkID),
+			zap.Stringer("status", choices.Accepted),
+		)
+		return
```

Comment:
Do we want to avoid writing all other metrics in this case?

Reply:
It should never happen 🤷 If this gets logged something really broken has happened